### PR TITLE
[VYMCA-104] Add Favorites option 

### PR DESCRIPTION
**Pull request by hamrant**
**Related Issue/Ticket:**

Basecamp - https://3.basecamp.com/3608022/buckets/17045978/todos/3182733185
Jira - https://fivejars.atlassian.net/browse/VYMCA-104

## Steps to test:

- [ ] Login to VY
- [ ] Go to `/virtual-ymca#/favorites`
- [ ] Check that you can see the message "There no favorite content."

Screenshot from 2020-12-07 18-06-34

- [ ] Click on the Adjust button and check that types options disabled

Screenshot from 2020-12-07 18-06-52

- [ ] Go to the home page
- [ ] Check that on listings you can add any content to favorites by click on the icon

Screenshot from 2020-12-03 18-19-56

- [ ] Go to any individual page and check that you can see favorites component

Screenshot from 2020-12-03 18-21-19

- [ ] Go to `/virtual-ymca#/favorites`
- [ ] Check that you can see all items added to favorites

Screenshot from 2020-12-07 18-10-50

- [ ] Click on the Adjust button and check that you can see modal with filters

Screenshot from 2020-12-07 18-11-43

- [ ] Check that content without favorites items disabled in the filter (this content hidden on the page)
- [ ] Test filters and sorting
- [ ] Check that by pressing the "view all" button you will see all items for specified content type
- [ ] Check that on page reload filters restored from data in URL

**Note** Styles will be fixed later by @retif

### CODE REVIEW

File: js/gated-content/src/components/AddToFavorite.vue [hunk 1]
```diff
@@ -0,0 +1,95 @@
+<template>
+  <div
+    :class="{ active: isActive, 'add-to-favorite': true }"
```
Comment (NightHunterSV) [inline js/gated-content/src/components/AddToFavorite.vue:3]: Shouldn't we move the 'add-to-favorite' a line below?

File: js/gated-content/src/components/event/EventListing.vue [hunk 2]
```diff
@@ -208,10 +235,7 @@ export default {
 
       params.filter.status = 1;
       params.sort = {
-        sortByDate: {
-          path: 'date.value',
-          direction: 'ASC',
-        },
+        sortByDate: this.sort,
```
Comment (NightHunterSV) [inline js/gated-content/src/components/event/EventListing.vue:238]: Let's move to `sortBy` for the consistency

File: js/gated-content/src/components/blog/BlogListing.vue [hunk 3]
```diff
@@ -67,6 +68,16 @@ export default {
       type: Boolean,
       default: false,
     },
+    favorites: {
```
Comment (NightHunterSV) [inline js/gated-content/src/components/blog/BlogListing.vue:71]: I suppose we can move this to a mixin also

File: js/gated-content/src/components/blog/BlogListing.vue [hunk 4]
```diff
@@ -132,6 +141,20 @@ export default {
         };
       }
 
+      if (this.favorites) {
```
Comment (NightHunterSV) [inline js/gated-content/src/components/blog/BlogListing.vue:144]: We can refactor this and move to the mixin

File: src/Controller/FavoritesController.php [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+    $data = $this->getData($request);
+    if ($data instanceof JsonResponse) {
+      return $data;
+    }
+    $favorite_items = $this->entityTypeManager()
+      ->getStorage('vy_favorite_item')
+      ->loadByProperties([
+        'uid' => $this->currentUser()->id(),
+        'ref_entity_type' => $data['entity_type'],
+        'ref_entity_bundle' => $data['entity_bundle'],
```
Comment (NightHunterSV) [inline src/Controller/FavoritesController.php:151]: Probably we can omit this. But I'm not insisting on this.

File: js/gated-content/src/components/AddToFavorite.vue (hunk 1 above)
Comment (hamrant) [reply js/gated-content/src/components/AddToFavorite.vue]: Fixed

File: js/gated-content/src/components/event/EventListing.vue (hunk 2 above)
Comment (hamrant) [reply js/gated-content/src/components/event/EventListing.vue]: Fixed

File: js/gated-content/src/components/blog/BlogListing.vue (hunk 3 above)
Comment (hamrant) [reply js/gated-content/src/components/blog/BlogListing.vue]: Fixed

File: js/gated-content/src/components/blog/BlogListing.vue (hunk 4 above)
Comment (hamrant) [reply js/gated-content/src/components/blog/BlogListing.vue:140]: Based on our conversation - decided to skip this.

File: src/Controller/FavoritesController.php (hunk 5 above)
Comment (hamrant) [reply src/Controller/FavoritesController.php:151]: Based on our conversation - decided to skip this.

**Review by NightHunterSV — APPROVED** [review]

**Review by NightHunterSV — APPROVED** [review]

File: src/VyFavoriteItemInterface.php [hunk 6]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+namespace Drupal\openy_gated_content;
+
+use Drupal\Core\Entity\ContentEntityInterface;
+use Drupal\user\EntityOwnerInterface;
+
+/**
+ * Provides an interface defining a path_alias entity.
```
Comment (anpolimus) [inline src/VyFavoriteItemInterface.php:9]: 1. Do we need this empty interface?
2. This is copied from path_alias

**Review by anpolimus — CHANGES_REQUESTED** [review]

File: src/VyFavoriteItemInterface.php (hunk 6 above)
Comment (hamrant) [reply src/VyFavoriteItemInterface.php:9]: I will fix comment and add methods that defined in the entity class

Comment (hamrant) [reply src/VyFavoriteItemInterface.php]: Fixed

**Comment by NadyaRudenko** [comment]
retest

**Review by anpolimus — APPROVED** [review]

**Pull request merged**